Vtol rework

tumbili opened the pull request:
- move tailsitter and tiltrotor to specific classes
- vtol attitude control module should only contain generic function calls but no airframe specific code

tumbili commented:
@LorenzMeier This is the best way I have found to integrate both changes from the FireFly and the Caipirinha into master. I don't think that parameters inside vtol_attitude_control module would have been nice, because the control of the two airframe types differs in many aspects.
My goal was to have a very simple structure in vtol_att_control_main which is very similar to the one before (have fw and mc in parallel). Then you have generic function calls to the airframe class where tailsitter/tiltorotor can do whatever they want to.

tumbili commented on src/modules/vtol_att_control/airframe.h:
```diff
@@ -0,0 +1,60 @@
+	virtual void update_external_state() = 0;
+};
+
+#endif
```
newline

tumbili commented on src/modules/vtol_att_control/tailsitter.cpp:
```diff
@@ -0,0 +1,162 @@
+	_actuators_out_1.control[0] = _actuators_mc_in.control[2];	//roll elevon
+	_actuators_out_1.control[1] = _actuators_mc_in.control[1];;	//pitch elevon
+	_actuators_out_1.control[4] = _tilt_control;	// for tilt-rotor control
+}
```
newline

tumbili commented on src/modules/vtol_att_control/tailsitter.h:
```diff
@@ -0,0 +1,65 @@
+		void scale_mc_output();
+	
+
+ };
```
newline

tumbili commented on src/modules/vtol_att_control/tiltrotor.cpp:
```diff
@@ -0,0 +1,175 @@
+	if (ret != OK) {errx(ret, "failed setting max values");}
+
+	close(fd);
+}
```
newline

tumbili commented on src/modules/vtol_att_control/tiltrotor.h:
```diff
@@ -0,0 +1,65 @@
+	void set_max_mc();
+	void set_max_fw(unsigned pwm_value);
+
+ };
```
newline

tumbili commented on src/modules/vtol_att_control/vtol_att_control_main.h:
```diff
@@ -0,0 +1,226 @@
+	void 		set_idle_mc();
+};
+
+#endif
```
newline

tumbili commented on src/modules/vtol_att_control/vtol_att_control_params.c:
```diff
@@ -142,3 +142,11 @@ PARAM_DEFINE_FLOAT(VT_PROP_EFF,0.0f);
+ * @max 1
+ * @group VTOL Attitude Control
+ */
+PARAM_DEFINE_INT32(VT_AIRFRAME, 0);
```
newline

LorenzMeier replied:
Looks like this is a type, not a specific airframe.

tumbili replied:
Yes, you are right.

LorenzMeier commented on ROMFS/px4fmu_common/init.d/13002_firefly6:
```diff
@@ -41,3 +41,4 @@ set MAV_TYPE 21
 
 param set VT_MOT_COUNT 6
 param set VT_IDLE_PWM_MC 1080
+param set VT_AIRFRAME 1
```
Newline missing.

LorenzMeier commented on ROMFS/px4fmu_common/init.d/13003_tailsitter_x:
```diff
@@ -0,0 +1,17 @@
+set MAV_TYPE 20
+param set VT_MOT_COUNT 4
+param set VT_IDLE_PWM_MC  1080
+param set VT_AIRFRAME 0
```
Newline missing.

LorenzMeier commented on ROMFS/px4fmu_common/mixers/tailsitter_x.main.mix:
```diff
@@ -0,0 +1,18 @@
+M: 2
+O:       10000   10000    0 -10000  10000
+S: 1 0   10000   10000    0 -10000  10000
+S: 1 1  -10000  -10000    0 -10000  10000
```
Newline missing.

LorenzMeier commented on src/modules/vtol_att_control/airframe.cpp:
```diff
@@ -0,0 +1,49 @@
+
+Airframe::~Airframe() {
+	
+}
```
Newline missing..

LorenzMeier commented on src/modules/vtol_att_control/tailsitter.h:
```diff
@@ -0,0 +1,65 @@
+#include "vtol_att_control_main.h"
+#include "airframe.h"
+
+ class Tailsitter : public VtolAttitudeControl, public Airframe
```
Multi-Inheritance is an evil pattern and generally better to be avoided - do you really need this here? Is the Tailsitter class an Airframe config or an Attitude controller? Why can't VTOLAttitudeControl just inherit from Airframe?

LorenzMeier commented on src/modules/vtol_att_control/vtol_att_control_main.cpp:
```diff
@@ -276,9 +119,22 @@ VtolAttitudeControl::VtolAttitudeControl() :
 	_params_handles.power_max = param_find("VT_POWER_MAX");
 	_params_handles.prop_eff = param_find("VT_PROP_EFF");
 	_params_handles.arsp_lp_gain = param_find("VT_ARSP_LP_GAIN");
+	_params_handles.airframe_type = param_find("VT_AIRFRAME");
```
Is the AIRFRAME here not rather encoding the type?

LorenzMeier commented:
Looks good in general. I would recommend you to fix small things like newlines just immediately yourself rather than spending the time to comment them - its tedious, but much faster.

The only architectural change that is needed is to drop the multiple inheritance. There are many potential problems and confusion associated to it, and I don't see why it would be needed here. Either have the attitude controller as a member of the VTOL airframe class, or just inherit from the Airframe class.
http://stackoverflow.com/questions/406081/why-should-i-avoid-multiple-inheritance-in-c

tumbili commented:
@LorenzMeier Yeah I understand. I can make the Airframe class inherit form VTOL_att_control and then let tailsitter and tiltrotor inherit from the airframe class.

tumbili commented:
@LorenzMeier 
- removed the double inheritance: I now have a vtol_type class which inherits from vtol_att_control class. Then the tailsitter and tilrotor class both inherit from the vtol_type class.
- renamed airframe class to vtol_type class and also changed the VT_AIRFRAME param to VT_TYPE

tumbili commented:
Closing in favour of #2418
